vtkext/private/module/vtkF3DRenderer.cxx

```diff
@@ -1548,6 +1640,109 @@ void vtkF3DRenderer::ResetCameraClippingRange()
+  {
+    for (size_t i = 0; i < 3; ++i)
+    {
+      size_t nonzerosI = 0;
+      size_t nonzerosJ = 0;
+      for (size_t j = 0; j < 3; ++j)
+      {
+        if (::abs(m->Element[i][j]) > tol)
+          nonzerosI++;
+        if (::abs(m->Element[j][i]) > tol)
+          nonzerosJ++;
+      }
+      if (nonzerosI > 1 || nonzerosJ > 1)
+        return false;
+    }
+    return true;
+  };
+
+  /* Use `PokeMatrix` around the call to `GetBounds()` to extend box.
+   * Only gives the thightest bounds if the transformation is axis-aligned. */
+  const auto extendBoxAxisAligned = [&](vtkProp3D* prop3d, vtkBoundingBox& box)
+  {
+    vtkNew<vtkMatrix4x4> tmpMatrix;
+    vtkMatrix4x4::Multiply4x4(matrix, prop3d->GetMatrix(), tmpMatrix);
+    prop3d->PokeMatrix(tmpMatrix);
+
+    box.AddBounds(prop3d->GetBounds());
+
+    prop3d->PokeMatrix(nullptr);
+  };
+
+  /* Use custom logic to extend box.
+   * Should give the tightest bounds even when non-axis-aligned */
+  const auto extendBoxArbitrary = [&](vtkProp3D* prop3d, vtkBoundingBox& box)
+  {
+    vtkActor* actor = dynamic_cast<vtkActor*>(prop3d);
+    if (actor)
+    {
+      vtkPolyDataMapper* polyMapper = dynamic_cast<vtkPolyDataMapper*>(actor->GetMapper());
+      if (polyMapper)
+      {
+        vtkNew<vtkMatrix4x4> tmpMatrix;
+        vtkMatrix4x4::Multiply4x4(matrix, actor->GetMatrix(), tmpMatrix);
+        vtkPolyData* polydata = polyMapper->GetInput();
+        if (polydata)
+        {
+          double p[4] = { 0, 0, 0, 1 };
+          double q[4];
+          for (vtkIdType i = 0; i < polydata->GetNumberOfPoints(); ++i)
+          {
+            polydata->GetPoint(i, p);
+            tmpMatrix->MultiplyPoint(p, q);
+            box.AddPoint(q);
+          }
+          return true;
+        }
+      }
+    }
+    return false;
+  };
+
+  const bool isAxisAligned = isMatrixAxisAligned(matrix);
+  vtkBoundingBox box;
+
+  /* use `ComputeVisiblePropBounds()`'s logic to iterate `vtkProp3D`s contributing to the bounds */
+  vtkProp* prop;
+  vtkCollectionSimpleIterator pit;
+  for (this->Props->InitTraversal(pit); (prop = this->Props->GetNextProp(pit));)
+  {
+    if (prop->GetVisibility() && prop->GetUseBounds())
+    {
+      const double* bounds = prop->GetBounds();
+      if (bounds != nullptr && vtkMath::AreBoundsInitialized(bounds))
+      {
+        vtkProp3D* prop3d = dynamic_cast<vtkProp3D*>(prop);
+        if (prop3d)
+        {
+          if (isAxisAligned)
+          {
+            extendBoxAxisAligned(prop3d, box);
+          }
+          else
+          {
+            if (!extendBoxArbitrary(prop3d, box))
+            {
+              const std::string repr = std::string(prop3d->GetClassName());
+              F3DLog::Print(F3DLog::Severity::Warning,
+                "Could not properly account for " + repr +
+                  " in non-axis-aligned bounds computation");
+              extendBoxAxisAligned(prop3d, box);
+            }
+          }
+        }
```

Comment:
I still think the bare minimum we can do is
```cpp
        if (prop3d) // we need an actual prop
        {
          if (isAxisAligned) // we want the fast code path if possible
          {
            extendBoxAxisAligned(prop3d, box);
          }
          else // otherwise we do the arbitrary computation
          {
            extendBoxArbitrary(prop3d, box); // but we don't check if it worked
          }
        }
```

but from there I don't think it costs much to handle the unlikely case of an unknown mapper and do
```cpp
        if (prop3d) // we need an actual prop
        {
          if (isAxisAligned) // we want the fast code path if possible
          {
            extendBoxAxisAligned(prop3d, box);
          }
          else if(!extendBoxArbitrary(prop3d, box)) // otherwise we try the arbitrary computation
          {
            extendBoxAxisAligned(prop3d, box); // the optimal arbitrary code didn't succeed so we fall back to the fast code that will work but give a sub-optimal (not tight) box
            // and we might as well add a log here I guess
          }
        }
```
which is basically the current version of the PR.

Am I missing something?

Reply:
Right